cpp/ql/lib/semmle/code/cpp/commons/Printf.qll

```diff
@@ -1225,35 +1225,30 @@ class FormatLiteral extends Literal {
     )
   }
 
-  private int getMaxConvertedLengthAfter(int n) {
+  private int getMaxConvertedLengthAfter(
+    int n, TFormattedSizeEstimateKind strategy, boolean relevantForStrategy
+  ) {
+    (relevantForStrategy = true or relevantForStrategy = false) and // avoid "is not bound"...
     if n = this.getNumConvSpec()
-    then result = this.getConstantSuffix().length() + 1
+    then
+      // for the moment, the Limited strat is always relevant
+      (strategy = TFormattedSizeLimited() or relevantForStrategy = true) and
+      result = this.getConstantSuffix().length() + 1
     else
-      result =
-        this.getConstantPart(n).length() + this.getMaxConvertedLength(n) +
-          this.getMaxConvertedLengthAfter(n + 1)
+      exists(boolean strategyRelevantForThisSpec |
+        result =
+          this.getConstantPart(n).length() +
+            this.getMaxConvertedLength(n, strategy, strategyRelevantForThisSpec) +
+            this.getMaxConvertedLengthAfter(n + 1, strategy,
+              relevantForStrategy.booleanOr(strategyRelevantForThisSpec))
+      )
   }
 
-  private int getMaxConvertedLengthAfterLimited(int n) {
-    if n = this.getNumConvSpec()
-    then result = this.getConstantSuffix().length() + 1
-    else
-      result =
-        this.getConstantPart(n).length() + this.getMaxConvertedLengthLimited(n) +
-          this.getMaxConvertedLengthAfterLimited(n + 1)
+  int getMaxConvertedLength(TFormattedSizeEstimateKind strategy) {
```

Comment:
This is part of the public interface, so it should have a QLDoc comment.  Also, we should preserve the original `getMaxConvertedLength()` and `getMaxConvertedLengthLimited()` predicates for backwards compatibility, even if all they do is call into the new one (and you can deprecate them if you like).